layer1/executor/task_executor.go

```diff
@@ -0,0 +1,354 @@
+				return false, nil
+			}
+
+			if receipt.Status == types.ReceiptStatusSuccessful {
+				logger.Debug("got a successful receipt")
+				return true, nil
+			} else {
+				logger.Warn("got a reverted receipt, retrying")
+				return false, nil
+			}
+		}
+
+		logger.Trace("receipt is not ready yet")
+		hasToExecute, err := shouldExecute(ctx, task)
+		if err != nil {
+			return false, err
+		}
+		if !hasToExecute {
+			return true, nil
+		}
+	}
+}
+
+// shouldExecute checks if a task should be executed. In case of recoverable errors the
+// function is going to retry `constants.MonitorRetryCount` times. If it
+// exhausted the number of retries, it sends true since there's no information.
+// The function returns false in case of unrecoverable errors, or if it
+// shouldn't execute a task. In case of no errors, the return value is true
+// (default case to t.ShouldExecute to return that a task should be executed).
+func shouldExecute(ctx context.Context, task tasks.Task) (bool, error) {
+	logger := task.GetLogger()
+	for i := uint64(1); i <= constants.MonitorRetryCount; i++ {
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		default:
+		}
+		if hasToExecute, err := task.ShouldExecute(ctx); err != nil {
+			if err.IsRecoverable() {
+				logger.Tracef("got a recoverable error during task.ShouldExecute: %v", err)
+				if err := sleepWithContext(ctx, constants.MonitorRetryDelay); err != nil {
+					return false, err
+				}
+				continue
+			}
+			logger.Debugf("got a non recoverable error during task.ShouldExecute: %v", err)
+			return false, err
+		} else {
+			logger.Tracef("should execute BaseTask: %v", hasToExecute)
+			return hasToExecute, nil
+		}
+	}
+
+	return true, nil
+}
+
+// persistState TaskExecutor to database.
+func (te *TaskExecutor) persistState() error {
+	logger := logging.GetLogger("staterecover").WithField("State", "task_executor")
+	rawData, err := json.Marshal(te)
+	if err != nil {
+		return err
+	}
+
+	err = te.database.Update(func(txn *badger.Txn) error {
+		key := dbprefix.PrefixTaskExecutorState()
+		logger.WithField("Key", string(key)).Debug("Saving state in the database")
+		if err := utils.SetValue(txn, key, rawData); err != nil {
+			logger.Error("Failed to set Value")
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := te.database.Sync(); err != nil {
+		logger.Error("Failed to set sync")
+		return err
+	}
+
+	return nil
+}
+
+// loadState TaskExecutor from database.
+func (te *TaskExecutor) loadState() error {
+	logger := logging.GetLogger("staterecover").WithField("State", "task_executor")
+	if err := te.database.View(func(txn *badger.Txn) error {
+		key := dbprefix.PrefixTaskExecutorState()
+		logger.WithField("Key", string(key)).Debug("Loading state from database")
+		rawData, err := utils.GetValue(txn, key)
+		if err != nil {
+			return err
+		}
+
+		err = json.Unmarshal(rawData, te)
+		if err != nil {
+			return err
```

Comment:
Cover in tests if possible. Should unwind main due to corrupt state.

Reply:
Sure thing, will cover it in tests. In theory the exit path should close the cf (Handler L:36) and panic on the validator.